qirlib/src/codegen/qis.rs

```diff
@@ -0,0 +1,227 @@
+    module: &Module<'ctx>,
+) -> FunctionValue<'ctx> {
+    get_rotated_intrinsic_function_body(context, module, "ry")
+}
+
+pub(crate) fn rz_body<'ctx>(
+    context: &'ctx inkwell::context::Context,
+    module: &Module<'ctx>,
+) -> FunctionValue<'ctx> {
+    get_rotated_intrinsic_function_body(context, module, "rz")
+}
+
+pub(crate) fn reset_body<'ctx>(
+    context: &'ctx inkwell::context::Context,
+    module: &Module<'ctx>,
+) -> FunctionValue<'ctx> {
+    get_intrinsic_function_body(context, module, "reset")
+}
+
+pub(crate) fn m_body<'ctx>(
+    context: &'ctx inkwell::context::Context,
+    module: &Module<'ctx>,
+) -> FunctionValue<'ctx> {
+    get_intrinsic_m_function_body(context, module, "m")
+}
+
+/// `declare void @__quantum__qis__{}__body(%Qubit*, %Qubit*)`
+pub(crate) fn get_controlled_intrinsic_function_body<'ctx>(
+    context: &'ctx inkwell::context::Context,
+    module: &Module<'ctx>,
+    name: &str,
+) -> FunctionValue<'ctx> {
+    let function_name = format!("__quantum__qis__{}__body", name.to_lowercase());
+    if let Some(function) = get_function(module, function_name.as_str()) {
+        function
+    } else {
+        let void_type = context.void_type();
+        let qubit_ptr_type = qubit(context, module).ptr_type(AddressSpace::Generic);
+        let fn_type = void_type.fn_type(&[qubit_ptr_type.into(), qubit_ptr_type.into()], false);
+        let fn_value =
+            module.add_function(function_name.as_str(), fn_type, Some(Linkage::External));
+        fn_value
+    }
+}
+
+/// `declare void @__quantum__qis__{}__body(double, %Qubit*)`
+pub(crate) fn get_rotated_intrinsic_function_body<'ctx>(
+    context: &'ctx inkwell::context::Context,
+    module: &Module<'ctx>,
+    name: &str,
+) -> FunctionValue<'ctx> {
+    let function_name = format!("__quantum__qis__{}__body", name.to_lowercase());
+    if let Some(function) = get_function(module, function_name.as_str()) {
+        function
+    } else {
+        let void_type = context.void_type();
+        let qubit_ptr_type = qubit(context, module).ptr_type(AddressSpace::Generic);
+        let fn_type = void_type.fn_type(&[double(context).into(), qubit_ptr_type.into()], false);
+        let fn_value =
+            module.add_function(function_name.as_str(), fn_type, Some(Linkage::External));
+        fn_value
+    }
+}
+
+/// `declare void @__quantum__qis__{}__body(%Qubit*)`
+pub(crate) fn get_intrinsic_function_body<'ctx>(
+    context: &'ctx inkwell::context::Context,
+    module: &Module<'ctx>,
+    name: &str,
+) -> FunctionValue<'ctx> {
+    let function_name = format!("__quantum__qis__{}__body", name.to_lowercase());
+    if let Some(function) = get_function(module, function_name.as_str()) {
+        function
+    } else {
+        let void_type = context.void_type();
+        let qubit_ptr_type = qubit(context, module).ptr_type(AddressSpace::Generic);
+        let fn_type = void_type.fn_type(&[qubit_ptr_type.into()], false);
```

Comment:
All of these void-typed intrinsic function getters follow very similar patterns where the only difference is the argument types. Could they be refactored to call a common base? The callers would look something like:
```rust
pub(crate) fn get_rotated_intrinsic_function_body<'ctx>(
    context: &'ctx inkwell::context::Context,
    module: &Module<'ctx>,
    name: &str,
) -> FunctionValue<'ctx> {
    let qubit_ptr_type = qubit(context, module).ptr_type(AddressSpace::Generic);
    get_intrinsic_function_body_helper(context, module, name, &[double(context).into(), qubit_ptr_type.into()])
}
```
That would mean the slice of parameter types gets constructed every time when it might not need to be, but perhaps that's a reasonable trade off to consider.